### PR TITLE
cli: niceify command invocvation and help

### LINKED ISSUE #2957: Minor check-software tweaks needed
Just discovered two problems using a Centos-based Docker container for Cylc

 - `cylc check-software` has an implicit dependence on the `which` command, which may not be installed by default (`yum install which` on RHEL systems).

We can't use `shutil.which()` till Python 3, but maybe `distutils.spawn.find_executable()`?

 - `cylc check-software` does an `import gtk` which evidently aborts with RuntimeError if no display is available - which it typically isn't in a container.

Is it possible to determine the right module version without needing a display?

### LINKED ISSUE #3208: The future of check-software
Following on from a point arising from reviewing https://github.com/cylc/cylc-doc/pull/26#issuecomment-510081950, after the:

1. splitting up of the ``cylc/cylc`` mono-repo to inter-connected multi-repositories; &
2. a number of large-scale package dependency changes from Cylc 7 to 8 (& beyond), so far notably the:
    * Python 2 -> 3 port;
    * docs conversion from LaTeX to Sphinx;
    * architectural changes;
    * development of the web UI,

we have completely changed the external software dependencies for our full ecosystem (see my outputs below), & divided them into subsets of requirements based on each sub-component.

**The ``check-software`` command currently sits under ``cylc-flow``, & reports what is needed for this. But its place going forward is into clear & merits discussion:**

1. Is it still worth having, noting that:
  * the ``setup.py`` of ``setuptools`` we now use has settings managing dependency specifications, both for [compulsory ones](https://setuptools.readthedocs.io/en/latest/setuptools.html?highlight=install_requires#declaring-dependencies) & [optional ones for subsidiary features](https://setuptools.readthedocs.io/en/latest/setuptools.html?highlight=install_requires#declaring-extras-optional-features-with-their-own-dependencies).
  * in the future we might even want, & be able to, provide ready-made environments e.g. ``conda`` via ``conda-forge`` that can be sourced, so that the user does not even have to consider dependencies & can just source those.
  * it may still be useful as a diagnostic tool, i.e. an admin command, if it is not useful to users.
2. What repository will it sit in?
3. What repositories should it capture dependencies for? Noting that:
  * users will generally not want to use every single (non-admin) component under the ``cylc`` organisation. For example, they are unlikely to have need to build the docs themselves, therefore ``sphinx`` as a docs requirement is not relevant to most users.
4. Should we amend it in any way, taking the above decisions into account?
  * How to organise under headings? Perhaps report on a repo-by-repo basis?
  * Most of the packages reported now are Python-based, so it seems sensible to drop the ``Python:`` descriptor prepended to each package in the output.

*************************************
#### Cylc ``8.0a0`` i.e. ``master`` (in ``cylc/cylc-flow``)
Outputs:
```
Checking your software...

Individual results:
================================================================================
Package (version requirements)                           Outcome (version found)
================================================================================
                              *REQUIRED SOFTWARE*                              
Python (3.7+)...........................FOUND & min. version MET (3.7.3.final.0)
Python:zmq (any)..................................................FOUND (18.0.0)
Python:jose (any)..................................................FOUND (3.0.1)
Python:graphene (any)..............................................FOUND (2.1.6)
Python:colorama (any)..............................................FOUND (0.4.1)
Python:ansimarkup (any)............................................FOUND (1.4.0)
Python:protobuf (3.7.0+)........................FOUND & min. version MET (3.7.1)

              *OPTIONAL SOFTWARE for the configuration templating*              
Python:EmPy (3.3.4+)...............................................NOT FOUND (-)
================================================================================

Summary:
                          ****************************                          
                             Core requirements: ok                             
                           Full-functionality: not ok                           
                          **************************** 
```

#### Cylc ``7.8.x`` (the old ``cylc/cylc``)
Outputs:
```
Checking your software...

Individual results:
================================================================================
Package (version requirements)                           Outcome (version found)
================================================================================
                              *REQUIRED SOFTWARE*                              
Python (2.6+, <3)......................FOUND & min. version MET (2.7.16.final.0)

        *OPTIONAL SOFTWARE for the GUI & dependency graph visualisation*        
Python:pygraphviz (any)............................................NOT FOUND (-)
graphviz (any)....................................................FOUND (2.30.1)
Python:pygtk (2.0+)................................................NOT FOUND (-)

             *OPTIONAL SOFTWARE for the HTTPS communications layer*             
Python:requests (2.4.2+)...........................................NOT FOUND (-)
Python:urllib3 (any)...............................................NOT FOUND (-)
Python:OpenSSL (any)...............................................NOT FOUND (-)

              *OPTIONAL SOFTWARE for the configuration templating*              
Python:EmPy (any)..................................................NOT FOUND (-)

                 *OPTIONAL SOFTWARE for the HTML documentation*                 
Python:sphinx (1.5.3+).............................................NOT FOUND (-)
================================================================================

Summary:
                          ****************************                          
                             Core requirements: ok                             
                           Full-functionality: not ok                           
                          **************************** 
```

### LINKED ISSUE #3488: cli: change how we re-invoke commands?
The `cylc` command currently re-invokes its sub-commands, e.g. `cylc run suite` becomes `cylc-run suite`.

https://github.com/cylc/cylc-flow/blob/8af55bc494d8b9904898262212f96b6806eea921/cylc/flow/scripts/cylc.py#L53

Now that the veil of the bash script has been stripped I'm not quite sure what the reason for this is.

* We can still easily grep for `cylc` processes.
* `cylc` can be easily white-listed.

Turns out `os.exec*` operations do not play with Code coverage (see #3486).

Questions:

* [x] Any objections to invoking `cylc` sub-commands via a Python interface without system re-invocation as `cylc-<sumcmd>`? **No**

**Pull requests welcome!**

### LINKED ISSUE #3893: cli: tidy
Having gone through the command list while working on the `rose suite-run` migration proposal, propose a quick tidy of commands and command help:

1) Hide all "internal" commands by default.
2) Drop the category system for invoking commands (i.e. the `task` in `cylc [task] message`).
3) Consider dropping the category system.
4) Replace `% cmd` code examples with `$ cmd` (because [bashism](https://en.wiktionary.org/wiki/bashism), but also [colour formatting](https://github.com/cylc/cylc-flow/pull/3680))
4) Remove the following commands in due course: (commented on #1249)
    * `check-software` - outdated and messy, setup.py ensures dependency compatibility.
    * `checkpoint` - #3891 
    * `client` - somewhat obsolete since GraphQL, remove or rename to `query` to match `subscribe`.
    * `get-directory` - will be made obsolete by #3889 
    * `ls-checkpoints` - #3891 
    * `register` - will be made obsolete by #3889 

Ping @hjoliver

### PR DESCRIPTION
**Pull request by oliver-sanders**
closes #3893 
closes #3488 
closes #3208 
closes #2957 
addresses #1249 
unblocks #3598 

@hjoliver Removed you are reviewer but make sure you are ok with the bullets below:

Highlights:

* Removes command re-invokation (i.e. `ps` now shows `cylc run` not `cylc-run`) [hence efficiency label]
* Remove `cylc check-software` - use standard tools like [pipdeptree](https://pypi.org/project/pipdeptree/)
* Reduce to two `console_scripts` entry points (`cylc`, `clyc`) so we only install top level commands.
* Created a `cylc.command` entry point and opened the interface to allow extensions to add Cylc sub-commands.
* Implement hidden aliases for removed commands inline with the [rose suite-run proposal](https://github.com/cylc/cylc-admin/blob/master/docs/proposal-rose-suite-run.md).
* Remove the command category system.
* Add a simple one line description of each command.
* Simplify the main `cylc` help page.
* Autogenerate the command listing (`cylc help all`)
* Remove the `cylc help` command (all functionalities purposefully removed or replaced)
* Splash of colour and a little TLC.

Before:

Screenshot 2020-10-27 at 18 43 28

After:

Screenshot 2020-10-29 at 19 01 52

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Does not need tests (why?).
- [x] Changelog included.
- [ ] Awaiting docs PR.
- [x] No dependency changes.

### CODE REVIEW

**Comment by hjoliver** [comment]
Nice, well done :tada:

**Comment by oliver-sanders** [comment]
Test chunk 2/4 is repeatably timing out, currently debugging by:

* Taking a copy of the prove output from the test run on GH actions > TESTS1
* CHUNK=2/4 etc/bin/run-functional-tests -dry > TESTS2
* Diff those files (with a bit of sed) to find out which tests haven't returned.

**Comment by oliver-sanders** [comment]
Rebased.

**Comment by oliver-sanders** [comment]
Removed a test which was defining its own `cylc run` client (incorrectly). Looking back at the test I'm not sure how valid it is, especially since the retry state was removed.

**Comment by hjoliver** [comment]
> @hjoliver Removed you are reviewer but make sure you are ok with the bullets below:

very happy :grin:

**Comment by hjoliver** [comment]
(small conflict now)

File: cylc/flow/scheduler_cli.py [hunk 1]
```diff
@@ -36,12 +37,15 @@
...
+from cylc.flow.scripts import cylc_header
 from cylc.flow import suite_files
 from cylc.flow.terminal import cli_function
 
 RUN_DOC = r"""cylc [control] run|start [OPTIONS] [ARGS]
 
-Start a suite run from scratch, ignoring dependence prior to the start point.
+Start a new suite run.
+
+This ignoring dependence prior to the start point.
```
Comment (datamel) [inline cylc/flow/scheduler_cli.py:48]:
```suggestion
Any dependencies prior to the start point will be ignored.
```

File: cylc/flow/scripts/cat_log.py [hunk 2]
```diff
@@ -39,14 +41,18 @@
 The correct cycle poin t format of the suite must be used for task job logs,
```
Comment (datamel) [inline cylc/flow/scripts/cat_log.py:41]:
```suggestion
The correct cycle point format of the suite must be used for task job logs,
```

File: cylc/flow/terminal.py [hunk 3]
```diff
@@ -55,6 +57,56 @@ def get_width(default=80):
         return default
 
 
+def centered(string, width=None):
+    """Print centeed text.
```
Comment (datamel) [inline cylc/flow/terminal.py:61]:
```suggestion
    """Print centered text.
```

**Review by datamel — APPROVED** [review]
Changes look great. 
I've checked out your branch and run commands. A couple of minor typos spotted while reading code.

**Review by wxtim — APPROVED** [review]

**Pull request merged**